### PR TITLE
RHCLOUD-30751 Principal Links endpoint fixes

### PR DESCRIPTION
**Pull request by EvanCasey13**
## Link(s) to Jira
https://issues.redhat.com/browse/RHCLOUD-30751

## Description of Intent of Change(s)
Fixed bug where incorrect links were being returned for the principals endpoint

## Local Testing
How can the feature be exercised?
How can the bug be exploited and fix confirmed?
Is any special local setup required?

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [ ] are there any pre/post merge actions required? if so, document here.
- [ ] are theses changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [ ] does this require migration changes?
  - [ ] if yes, are they backwards compatible?
- [ ] is there known, direct impact to dependent teams/components?
  - [ ] if yes, how will this be handled?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: rbac/management/principal/view.py [hunk 1]
```diff
@@ -188,7 +187,7 @@ def get(self, request):
                 "first": f"{path}?limit={limit}&offset=0{usernames_filter}",
                 "next": f"{path}?limit={limit}&offset={offset + limit}{usernames_filter}",
                 "previous": f"{path}?limit={limit}&offset={previous_offset}{usernames_filter}",
-                "last": None,
+                "last": f"{path}?limit={limit}&offset={int(limit - int(count))}{usernames_filter}",
```
Comment (petracihalova) [inline rbac/management/principal/view.py:190]: if limit < count then offset will be negative

maybe we can reuse the `get_paginated_response()` from [rbac/api/common/pagination.py](https://github.com/RedHatInsights/insights-rbac/blob/af853082447418d58a1cdde5e8bbdf350f8ba0ae/rbac/api/common/pagination.py#L80)

File: rbac/management/principal/view.py [hunk 2]
```diff
@@ -183,12 +182,13 @@ def get(self, request):
                 count = len(data)
             else:
                 count = None
+            last_link = int(count) - int(limit) if (int(count) - int(limit)) >= 0 else 0
```
Comment (petracihalova) [inline rbac/management/principal/view.py:185]: please consider better name for the variable, it should describe the value that is inside

File: rbac/management/principal/view.py [hunk 3]
```diff
@@ -183,12 +182,13 @@ def get(self, request):
...
             response_data["meta"] = {"count": count, "limit": limit, "offset": offset}
             response_data["links"] = {
                 "first": f"{path}?limit={limit}&offset=0{usernames_filter}",
                 "next": f"{path}?limit={limit}&offset={offset + limit}{usernames_filter}",
                 "previous": f"{path}?limit={limit}&offset={previous_offset}{usernames_filter}",
-                "last": None,
+                "last": f"{path}?limit={limit}&offset={last_link}{usernames_filter}",
```
Comment (petracihalova) [inline rbac/management/principal/view.py:191]: it is needed to fix all links here to work in same way as for other endpoints like `GET /groups/` for example `GET /principals/` for limit=10 and offset=0 (default values) the previous link should be `null`

File: rbac/management/principal/view.py (hunk 1 above)
Comment (lpichler) [reply rbac/management/principal/view.py]: @EvanCasey13 is it possible to leverage `get_paginated_response ` somehow ?
it looks there some cases which needs to covered.

File: rbac/management/principal/view.py [hunk 4]
```diff
@@ -123,7 +123,8 @@ def get(self, request):
         previous_offset = 0
         if offset - limit > 0:
             previous_offset = offset - limit
-
+        else:
+            None
```
Comment (astrozzc) [inline rbac/management/principal/view.py:127]: This seems can be removed

File: rbac/management/principal/view.py [hunk 5]
```diff
@@ -183,12 +184,15 @@ def get(self, request):
                 count = len(data)
             else:
                 count = None
+            last_link_offset = int(count) - int(limit) if (int(count) - int(limit)) >= 0 else 0
```
Comment (astrozzc) [inline rbac/management/principal/view.py:187]: Nit: we can put int(count) - int(limit) into a variable so we don't have to calculate it twice

**Comment by astrozzc** [comment]
Oh, as Libor mentioned, can we reuse the get_paginated_response to achieve this to simplify it. An example usage is here: https://github.com/RedHatInsights/insights-rbac/blob/master/rbac/management/access/view.py#L130-L132

**Pull request closed without merging**